jdbc/src/java/org/apache/hive/jdbc/saml/HiveJdbcBrowserClient.java

```diff
@@ -0,0 +1,322 @@
+import org.apache.hive.service.auth.saml.HiveSamlUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is used to execute a browser based SSO workflow with the authentication mode
+ * is browser.
+ */
+public class HiveJdbcBrowserClient implements IJdbcBrowserClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HiveJdbcBrowserClient.class);
+  // error message when the socket times out.
+  @VisibleForTesting
+  public static final String TIMEOUT_ERROR_MSG = "Timed out while waiting for server response";
+  private final ServerSocket serverSocket;
+  private HiveJdbcBrowserServerResponse serverResponse;
+  protected JdbcBrowserClientContext clientContext;
+  // By default we wait for 2 min unless overridden by a JDBC connection param
+  // browserResponseTimeout
+  private static final int DEFAULT_SOCKET_TIMEOUT_SECS = 120;
+  private final ExecutorService serverResponseThread = Executors.newSingleThreadExecutor(
+      new ThreadFactoryBuilder().setNameFormat("Hive-Jdbc-Browser-Client-%d")
+          .setDaemon(true).build());
+
+  HiveJdbcBrowserClient(JdbcConnectionParams connectionParams)
+      throws HiveJdbcBrowserException {
+    serverSocket = getServerSocket(connectionParams.getSessionVars());
+  }
+
+  private ServerSocket getServerSocket(Map<String, String> sessionConf)
+      throws HiveJdbcBrowserException {
+    final ServerSocket serverSocket;
+    int port = Integer.parseInt(sessionConf
+        .getOrDefault(JdbcConnectionParams.AUTH_BROWSER_RESPONSE_PORT, "0"));
+    int timeout = Integer.parseInt(
+        sessionConf.getOrDefault(JdbcConnectionParams.AUTH_BROWSER_RESPONSE_TIMEOUT_SECS,
+            String.valueOf(DEFAULT_SOCKET_TIMEOUT_SECS)));
+    try {
+      serverSocket = new ServerSocket(port, 0,
+          InetAddress.getByName(HiveSamlUtils.LOOP_BACK_INTERFACE));
+      LOG.debug("Browser response timeout is set to {} seconds", timeout);
+      serverSocket.setSoTimeout(timeout * 1000);
+    } catch (IOException e) {
+      throw new HiveJdbcBrowserException("Unable to bind to the localhost");
+    }
+    return serverSocket;
+  }
+
+  public Integer getPort() {
+    return serverSocket.getLocalPort();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (serverSocket != null) {
+      serverSocket.close();
+    }
+  }
+
+  public void init(JdbcBrowserClientContext clientContext) {
+    // everytime we set the sso URI we should clean up the previous state if its set.
+    // this may be from the previous invalid connection attempt or if the token has
+    // expired
+    reset();
+    this.clientContext = clientContext;
+    LOG.trace("Initialized the JDBCBrowser client with URL {}",
+        clientContext.getSsoUri());
+  }
+
+  private void reset() {
+    serverResponse = null;
+    clientContext = null;
+  }
+
+  private boolean validateSSOUrl(URI ssoUrl) {
+    //TODO(Vihang) add URL validation code here
+    return true;
```

Comment:
looks like this needs some real validation code?

Reply:
Initially I thought we can add a validation check to confirm that the SSO URL is from an expected domain. However, I plan to move that to a separate task since it is not critical. I removed this from the patch.